source/slang/slang-check-decl.cpp

```diff
@@ -23,6 +23,221 @@
+            }
+            continue;
+        }
+    }
+}
+
+static InterfaceDecl* findDynInheritance(SemanticsDeclVisitorBase* visitor, AggTypeDeclBase* decl)
+{
+    for (auto m : decl->members)
+    {
+        auto inheritanceDecl = as<InheritanceDecl>(m);
+        if (!inheritanceDecl)
+            continue;
+        visitor->ensureDecl(inheritanceDecl, DeclCheckState::ReadyForLookup);
+        auto inheritedInterfaceDeclRefType = as<DeclRefType>(inheritanceDecl->base.type);
+        if (!inheritedInterfaceDeclRefType)
+            continue;
+
+        auto inheritedInterfaceDecl =
+            as<InterfaceDecl>(inheritedInterfaceDeclRefType->getDeclRef().getDecl());
+        if (!inheritedInterfaceDecl)
+            continue;
+
+        if (!inheritedInterfaceDecl->hasModifier<DynModifier>())
+            continue;
+
+        return inheritedInterfaceDecl;
+    }
+    return nullptr;
+}
+static void validateDynUsage(
+    SemanticsDeclVisitorBase* visitor,
+    DiagnosticSink* sink,
+    CompilerOptionSet& optionSet,
+    Decl* decl)
+{
+    if (auto aggTypeDecl = as<AggTypeDecl>(decl))
+    {
+        // Ensure if we inherit from a `dyn interface` that we do not have: opaque types, unsized
+        // types, non-copyable types
+        auto inheritedDynInterface = findDynInheritance(visitor, aggTypeDecl);
+        if (inheritedDynInterface)
+        {
+            for (auto m : aggTypeDecl->members)
+            {
+                auto varDecl = as<VarDecl>(m);
+                if (!varDecl)
+                    continue;
+
+                visitor->ensureDecl(varDecl, DeclCheckState::ReadyForLookup);
+
+                if (isNonCopyableType(varDecl->getType()))
+                {
+                    sink->diagnose(
+                        m,
+                        Diagnostics::cannotHaveNonCopyableMemberWhenInheritingDynInterface,
+                        m,
+                        inheritedDynInterface);
+                }
+
+                int varTypeTags = (int)visitor->getTypeTags(varDecl->getType());
+                bool isUnsized = varTypeTags & (int)TypeTag::Unsized;
+                bool isOpaque = varTypeTags & (int)TypeTag::Opaque;
+                if (isUnsized)
+                    sink->diagnose(
+                        m,
+                        Diagnostics::cannotHaveUnsizedMemberWhenInheritingDynInterface,
+                        m,
+                        inheritedDynInterface);
+                if (isOpaque)
+                    sink->diagnose(
+                        m,
+                        Diagnostics::cannotHaveOpaqueMemberWhenInheritingDynInterface,
+                        m,
+                        inheritedDynInterface);
+            }
+        }
+
+        // validate simple `dyn interface` rules
+        auto interfaceDecl = as<InterfaceDecl>(decl);
+        if (!interfaceDecl)
+            return;
+        validateDynInterfaceUsage(visitor, sink, optionSet, interfaceDecl);
```

Comment:
We shouldn't need to call this from `validateDynUsage`, instead `validateDynInterfaceUsage` should be called from `visitInterfaceDecl` when we check the interface decl itself.